rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,527 @@
+   to help administrators resolve migration bugs.
+5. Educate Kibana developers to no longer simply write back an unmigrated
+   document if an exception occurred. A migration function should either
+   successfully transform the object or throw. 
+
+## 4.3 Automatically retry failed migrations until they succeed
+
+> Achieves goals: (2.2)
+> Mitigates errors (3.3) and (3.4)
+
+External conditions such as failures from an unhealthy Elasticsearch cluster
+(3.3) can cause the migration to fail. The Kibana cluster should be able to
+recover automatically once these external conditions are resolved. There are
+two broad approaches to solving this problem based on whether or not
+migrations are idempotent: 
+
+| Idempotent migrations |Description                                                |
+| --------------------- | --------------------------------------------------------- |
+| Yes                   | Idempotent migrations performed without coordination      |
+| No                    | Single node migrations coordinated through a lease / lock |
+
+Idempotent migrations don't require coordination making the algorithm
+significantly less complex and will never require manual intervention to
+retry. We, therefore, prefer this solution, even though it introduces
+restrictions on migrations (4.3.1.1). For other alternatives that were
+considered see section [(5)](#5-alternatives).
+
+## 4.3.1 Idempotent migrations performed without coordination
+
+The migration system can be said to be idempotent if the same results are
+produced whether the migration was run once or multiple times. This property
+should hold even if new (up to date) writes occur in between migration runs
+which introduces the following restrictions:
+
+### 4.3.1.1 Restrictions
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+### 4.3.1.2 Algorithm
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed. 
+2. Create `.kibana_n+1`, if it already exists, ignore the error and continue.
+3. Migrate documents by reading from `.kibana_n` and creating documents in
+   `.kibana_n+1`. If a document already exists, ignore.
```

Comment:
Is the migration author supposed to catch and ignore these errors or the migration system? Is there any concern that ignoring this could hide legit errors with id conflicts since users can manually set ids using json imports?

Reply:
The migration system will catch and handle any errors. We have some existing migrations which catch exceptions and return documents without transforming them, these should be updated to throw these errors.

While we run migration transform functions on imported documents before creating/updating them this doesn't cause a kibana-wide saved object migration. Our current terminology isn't very precise, but we could say that data is always transformed when imported, but we only run a migration when Kibana is upgraded and it finds outdated documents in the saved object indices.

There is a risk here if a migration transformation function changes id's or creates new documents (something we might support in the future). If there is an ID conflict it would go unnoticed leading to silent data loss. The migration to support sharing objects across multiple namespaces will generate new id's, but because it will use UUIDv5 it cannot conflict with any UUIDv4 ID's. Data loss is still theoretically possible if a user created an object with a custom UUIDv5 id that happens to match. I think the risk is small enough to be able to ignore.

We can't prevent this data loss, but we could at least detect it and fail the migration if this happens. To do this we could keep a list of all failed updates because a document already exists. Then once the migration is complete we check if any of the failed documents have an outdated migration version, if they do we fail the migration.